### ISSUE TITLE
Add support for abstract step sizes and dyads

### ISSUE THREAD

**Issue by inthar-raven** [xenharmonic-devs/sonic-weave#165]
Given step sizes `x_1, ..., x_n` as variables, _abstract dyads_ are linear combinations of step sizes `a_1 x_1 + ... + a_n x_n`. One use case could be how the intervals are displayed in the Analysis tab of Scale Workshop for scales (most useful for scales using up to 4 step sizes).

**Comment by frostburn** [comment]
I fail to grok the concept. Can you give an example?

**Comment by inthar-raven** [comment]
For a scale of the form:
```js
let x = 7/6
let y = 8/7
let z = 9/8
x; y; x; y; z;
stack()
```
the interval matrix with that setting would look like:
```
     |  1  |  2  |  3  |  4   |  (5)  | 
---------------------------------------
0    |  x  | x+y |2x+y |2x+2y |2x+2y+z|
x    |  y  | x+y |x+2y |x+2y+z|2x+2y+z|
x+y  |  x  | x+y |x+y+z|2x+y+z|2x+2y+z|
2x+y |  y  | y+z |x+y+z|x+2y+z|2x+2y+z|
2x+2y|  z  | x+z |x+y+z|2x+y+z|2x+2y+z|
```

**Comment by arseniiv** [comment]
Yeah looks pretty symbol-y, though the concepts of variables and symbols can be disjointed. How about something like:
```js
let x = 7/6 as 'L'
let y = 8/7 as 'M'
...
```
with `7/6 as 'L'` putting `7/6` into a box named `'L'`, `x` remaining a regular variable. Then these boxes can form linear combinations. In fact boxes with a single value aren’t needed and the new entities might just end up like JI-subgroup monzos, likewise belonging to logarithmic domain, so actually `8/7 as 'M'` will give out something like `Abstract({'dictionary': {'M': 8/7}, 'count': {'M': 1}})` and these can be added only if their dictionaries are coherent, having all equal step names be equal to the same concrete interval.

The main difficulty would be in adding another simplification function that evaluates the abstract expression, or making one of the existing ones fit to this role. Well and also doing something with many things that are expected of an interval. Well yeah there would be many difficulties…

---

What occurred to me instead right now is that one can try to “unparse” a concrete interval and add a name to it that corresponds to its monzo in a given subgroup (but written as a linear combination of step names), something like `unparse(7/6.8/7.9/8, 'x', 'y', 'z')` which would be also applicable elementwise to the whole scale at the end of its construction. Well the naming is bad, maybe something like `annotateStepCounts`.

**Comment by frostburn** [comment]
This is basically about solving a system of linear equations in logarithmic space. The variable names are technically available in the `StatementVisitor` instance, but I haven't decided if that API is going to stay public as we move to towards the beta.

**Comment by arseniiv** [comment]
I don’t think variable names should matter, too, though it *will* make some use cases too wordy. Like @inthar-raven’s example will become:
```js
let x = 7/6
let y = 8/7
let z = 9/8
x; y; x; y; z;
stack()
$ = map(annotateStepCounts(7/6.8/7.9/8, 'x', 'y', 'z'), $) // forgot how to write mapping on $ shorter
// also is it valid code to write `x.y.z` or it should be literals?
```

**Comment by frostburn** [comment]
Following the discussion at https://github.com/xenharmonic-devs/scale-workshop/pull/622 I think all `sonic-weave` needs to do is to turn Zarlino[10] into the string "mMLsLMmMLm". The analysis and manipulation of that string fits better with `moment-of-symmetry`.

**Comment by inthar-raven** [comment]
Could that package could be renamed to `scale-theory`, then?

**Comment by inthar-raven** [comment]
> Following the discussion at [xenharmonic-devs/scale-workshop#622](https://github.com/xenharmonic-devs/scale-workshop/pull/622) I think all `sonic-weave` needs to do is to turn Zarlino[10] into the string "mMLsLMmMLm". The analysis and manipulation of that string fits better with `moment-of-symmetry`.

Could that package be renamed to `scale-theory`?

**Comment by frostburn** [comment]
It's already on npm so I'm against renaming:
https://www.npmjs.com/package/moment-of-symmetry

Another home for abstract methods could be `xen-dev-utils`.

**Comment by frostburn** [comment]
Or here, if we must. SonicWeave is very much about scales, so getting abstract is not a huge leap.

**Comment by inthar-raven** [comment]
Should I create a new .ts file for it? `word.ts` or something.

**Comment by frostburn** [comment]
> Should I create a new .ts file for it? `word.ts` or something.

I anticipate multiple abstract utilities so `words.ts` should fit better.
